src/coreclr/jit/helperexpansion.cpp

```diff
@@ -2031,28 +2238,41 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
 
     // Block 2: typeCheckBb
     // TODO-InlineCast: if likelyCls == expectedCls we can consider saving to a local to re-use.
-    GenTree* likelyClsNode = gtNewIconEmbClsHndNode(likelyCls);
+    GenTree* likelyClsNode = gtNewIconEmbClsHndNode(expectedExactCls);
     GenTree* mtCheck       = gtNewOperNode(GT_EQ, TYP_INT, gtNewMethodTableLookup(gtCloneExpr(tmpNode)), likelyClsNode);
     mtCheck->gtFlags |= GTF_RELOP_JMP_USED;
     GenTree*    jtrue       = gtNewOperNode(GT_JTRUE, TYP_VOID, mtCheck);
     BasicBlock* typeCheckBb = fgNewBBFromTreeAfter(BBJ_COND, nullcheckBb, jtrue, debugInfo, lastBb, true);
 
     // Block 3: fallbackBb
-    GenTree*    fallbackTree = gtNewTempStore(tmpNum, call);
-    BasicBlock* fallbackBb   = fgNewBBFromTreeAfter(BBJ_ALWAYS, typeCheckBb, fallbackTree, debugInfo, lastBb, true);
+    BasicBlock* fallbackBb;
+    if (typeCheckFailedAction == F_CallHelper_AlwaysThrows)
+    {
+        // fallback call is used only to throw InvalidCastException
+        fallbackBb = fgNewBBFromTreeAfter(BBJ_THROW, typeCheckBb, gtUnusedValNode(call), debugInfo, nullptr, true);
+    }
+    else if (typeCheckFailedAction == F_ReturnNull)
+    {
+        // if fallback call is not needed, we just assign null to tmp
+        GenTree* fallbackTree = gtNewTempStore(tmpNum, gtNewNull());
+        fallbackBb            = fgNewBBFromTreeAfter(BBJ_ALWAYS, typeCheckBb, fallbackTree, debugInfo, lastBb, true);
+    }
+    else
+    {
+        GenTree* fallbackTree = gtNewTempStore(tmpNum, call);
+        fallbackBb            = fgNewBBFromTreeAfter(BBJ_ALWAYS, typeCheckBb, fallbackTree, debugInfo, lastBb, true);
+    }
```

Comment:
It seems the nice comment above is no longer fully accurate since the fallback BB can also assign null in some cases now.

Reply:
Addressed